src/Assert.php

```diff
@@ -309,12 +309,13 @@ public static function isInstanceOf(mixed $value, string $classString): object
     /**
      * Assert value is contained within the haystack. Strict comparison is used.
      * @template T
-     * @template R
+     * @template R of array
+     * @phpstan-assert T&value-of<R> $value
      *
      * @param T                    $value
-     * @param array<int|string, R> $haystack
+     * @param R                    $haystack
      *
-     * @return T&R
+     * @return T&value-of<R>
```

Comment:
Is `T&value-of<R>` so this will actually narrow the type instead of broaden it when `R` contains more types than `T`.